src/Microsoft.Azure.WebJobs.Host/Indexers/FunctionIndexer.cs

```diff
@@ -351,23 +343,72 @@ private bool HasJobAttribute(CustomAttributeData attributeData)
             };
         }
 
-        private class ListenerFactory : IListenerFactory
+        internal class ListenerFactory : IListenerFactory
         {
             private readonly FunctionDescriptor _descriptor;
             private readonly ITriggeredFunctionExecutor _executor;
             private readonly ITriggerBinding _binding;
+            private readonly TraceWriter _trace;
 
-            public ListenerFactory(FunctionDescriptor descriptor, ITriggeredFunctionExecutor executor, ITriggerBinding binding)
+            public ListenerFactory(FunctionDescriptor descriptor, ITriggeredFunctionExecutor executor, ITriggerBinding binding, TraceWriter trace)
             {
                 _descriptor = descriptor;
                 _executor = executor;
                 _binding = binding;
+                _trace = trace;
             }
 
-            public Task<IListener> CreateAsync(CancellationToken cancellationToken)
+            public async Task<IListener> CreateAsync(CancellationToken cancellationToken)
             {
                 ListenerFactoryContext context = new ListenerFactoryContext(_descriptor, _executor, cancellationToken);
-                return _binding.CreateListenerAsync(context);
+                IListener listener = await _binding.CreateListenerAsync(context);
+                return new TriggerListener(listener, _descriptor, _trace);
+            }
+
+            private class TriggerListener : IListener
+            {
+                private readonly IListener _listener;
+                private readonly FunctionDescriptor _descriptor;
+                private readonly TraceWriter _trace;
+
+                public TriggerListener(IListener listener, FunctionDescriptor descriptor, TraceWriter trace)
+                {
+                    _listener = listener;
+                    _descriptor = descriptor;
+                    _trace = trace;
+                }
+
+                public void Cancel()
+                {
+                    _listener.Cancel();
+                }
+
+                public void Dispose()
+                {
+                    _listener.Dispose();
+                }
+
+                public async Task StartAsync(CancellationToken cancellationToken)
+                {
+                    try
+                    {
+                        await _listener.StartAsync(cancellationToken);
+                    }
+                    catch (Exception e)
+                    {
+                        var fex = new FunctionIndexingException(_descriptor.ShortName, e);
```

Comment:
Yeah, I initially suggested that he piggyback on FEX since we already use that for similar purposes. I did also suggest the possibility of creating a base class that has the "handle" capability, e.g. HandleableException, with our existing FunctionIndexingException and a new FunctionListenerException deriving from it. Not sure the reuse of FEX is that big of an issue really - I can see thinking of listener startup as an index time thing in a loose way.

However, since I had these thoughts before, and Mike did also, lets just put the base class in place.